packages/core/src/api/blockManipulation/selections/selection.ts

```diff
@@ -127,19 +129,84 @@ export function getSelection<
     );
   }
 
+  const location = resolvePMToLocation(tr.doc, {
+    anchor: tr.selection.anchor,
+    head: tr.selection.head,
+  });
+
   return {
+    meta: {
+      location,
+      anchor: tr.selection.anchor,
+      head: tr.selection.head,
+      from: tr.selection.from,
+      to: tr.selection.to,
+    },
+    range: normalizeToRange(location),
     blocks,
+    /**
+     * The content of the selection, but in a way that it is always a valid slice of blocks.
+     */
+    get content() {
+      /**
+       * This implements the logic to "cut" a block from the start & end of the selection (no matter where it is)
+       * And return that content as a slice of blocks. The start & end are the only ones that would be different from the blocks array.
+       */
+      const endNode = withinSingleBlock
+        ? startNode
+        : indexToNode(endIndex, sharedDepth);
+
+      let cutStartNode: Node | undefined;
+      let cutEndNode: Node | undefined;
+
+      // Find the start and end nodes within the selection content
+      getSelectionContent(tr).content.descendants((node) => {
+        // Search for the start node by type and id
+        if (
+          !cutStartNode &&
+          node.type.name === startNode.type.name &&
+          node.attrs.id === startNode.attrs.id
+        ) {
+          cutStartNode = node;
+        }
+        // Search for the end node by type and id
+        if (
+          !cutEndNode &&
+          node.type.name === endNode.type.name &&
+          node.attrs.id === endNode.attrs.id
+        ) {
+          cutEndNode = node;
+        }
+        return !cutStartNode || !cutEndNode;
+      });
+
+      // If we didn't find the start or end node, return no blocks
+      if (!cutStartNode || !cutEndNode) {
+        return [];
+      }
+
+      // If selection is within a single node, just return that node
+      if (cutStartNode === cutEndNode) {
+        return [nodeToBlock(cutStartNode)];
+      }
+
+      // Return start node, middle nodes, and end node
+      return [
+        nodeToBlock(cutStartNode),
+        ...blocks.slice(1, -1),
+        nodeToBlock(cutEndNode),
+      ];
+    },
   };
 }
 
 export function setSelection(
```

Comment:
this function only works for blocks right? not points / ranges?

Reply:
This one is the block selection, the text setTextCursorPosition one handles points/ranges. I would prefer to only have one selection type, but I need an answer to https://github.com/TypeCellOS/BlockNote/pull/2044/files/ea043cd944e23cd3b52166c039eafedef1af2535#r2394892594 before I can merge them both into 1 selection interface.

I feel like you should always be able to get the current selection whether it is across multiple blocks or just a single collapsed selection within a block. But, I'm wanting to check how load-bearing this is first.